File: src/main/java/gruvexp/bbminigames/twtClassic/avatar/NpcAvatar.java
package gruvexp.bbminigames.twtClassic.avatar;

import gruvexp.bbminigames.Main;
import gruvexp.bbminigames.twtClassic.BotBows;
import gruvexp.bbminigames.twtClassic.BotBowsPlayer;
import gruvexp.bbminigames.twtClassic.hazard.HazardType;
import net.kyori.adventure.bossbar.BossBar;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.text.format.TextDecoration;
import net.kyori.adventure.title.Title;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Mannequin;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.LeatherArmorMeta;
import org.bukkit.inventory.meta.SkullMeta;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.potion.PotionEffect;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.Objects;
import java.util.Set;
import java.util.UUID;

public class NpcAvatar implements BotBowsAvatar{

    private Mannequin mannequin;
    private final BotBowsPlayer bp;
    private int visualHp = -1;

    public NpcAvatar(Mannequin mannequin, BotBowsPlayer bp) {
        this.mannequin = mannequin;
        this.bp = bp;
    }


    @Override
    public void message(Component component) {

    }

    @Override
    public LivingEntity getEntity() {
        return mannequin;
    }

    @Override
    public BotBowsPlayer getBotBowsPlayer() {
        return bp;
    }

    @Override
    public void eliminate() {
        Mannequin newMannequin = (Mannequin) Main.WORLD.spawnEntity(bp.getTeam().tribunePos, EntityType.MANNEQUIN);
        newMannequin.setProfile(mannequin.getProfile());
        newMannequin.customName(mannequin.customName());
        newMannequin.setCustomNameVisible(true);
        UUID oldId = getUUID();
        UUID newId = newMannequin.getUniqueId();
        BotBows.replacePlayerId(oldId, newId);
        mannequin.setHealth(0);
        mannequin = newMannequin;
        setInvis(true);
    }

    @Override
    public void revive() {
        setInvis(false);
    }

    @Override
    public void setHP(int hp) {
        visualHp = hp;
        if (hp == 0) {
            eliminate();
        } else {
            updateArmor();
        }
    }

    @Override
    public void setMaxHP(int maxHP) {
        setHP(maxHP);
    }

    @Override
    public ArmorSet getArmor() {
        ItemStack[] armor = mannequin.getEquipment().getArmorContents();
        return new ArmorSet(armor[0], armor[1], armor[2], armor[3]);
    }

    @Override
    public void remove() {
        mannequin.remove();
    }

    @Override
    public void reset() {

    }

    @Override
    public void readyBattle() {

    }

    @Override
    public void setReady(boolean ready, int itemIndex) {

    }

    @Override
    public int getNextFreeSlot() {
        return 0;
    }

    @Override
    public void damage() {
        mannequin.damage(0.001);
        mannequin.setGlowing(true);
        mannequin.setInvulnerable(true);

        Bukkit.getScheduler().runTaskLater(Main.getPlugin(), () -> {
            mannequin.setGlowing(false);
            mannequin.setInvulnerable(false);
        }, BotBows.HIT_DISABLED_ITEM_TICKS);
    }

    @Override
    public void setGlowing(boolean flag) {
        mannequin.setGlowing(flag);
    }

    @Override
    public void addPotionEffect(PotionEffect effect) {
        mannequin.addPotionEffect(effect);
    }

    @Override
    public void setColor(TextColor color) {

    }

    @Override
    public void growSize(double scale, int duration, int delay) {
        new BukkitRunnable() {
            int i = 1;
            final double scale0 = getRequiredAttribute(Attribute.SCALE).getBaseValue();
            @Override
            public void run() {
                if (i == duration) {
                    this.cancel();
                }
                getRequiredAttribute(Attribute.SCALE).setBaseValue(scale0 + (scale - scale0)/duration * i);
                i++;
            }
        }.runTaskTimer(Main.getPlugin(), delay, 1L);
    }

    @Override
    public UUID getUUID() {
        return mannequin.getUniqueId();
    }

    @Override
    public boolean isSneaking() {
        return mannequin.isSneaking();
    }

    @Override
    public void updateSneakStamina(float progress) {
        if (progress >= 1) mannequin.setSneaking(false);
    }

    @Override
    public ItemStack getHeadItem() {
        ItemStack item = new ItemStack(Material.PLAYER_HEAD);
        SkullMeta itemMeta = (SkullMeta) item.getItemMeta();
        itemMeta.displayName(bp.getName().decoration(TextDecoration.ITALIC, false));
        NamespacedKey key = new NamespacedKey(Main.getPlugin(), "uuid");
        itemMeta.getPersistentDataContainer().set(key, PersistentDataType.STRING, mannequin.getUniqueId().toString());
        itemMeta.setOwningPlayer(Bukkit.getOfflinePlayer("Robotagz"));

        item.setItemMeta(itemMeta);
        return item;
    }

    @Override
    public void setItem(int index, ItemStack item) {

    }

    @Override
    public void setInvis(boolean invis) {
        if (invis) {
            mannequin.getEquipment().setArmorContents(new ItemStack[] {null, null, null, null});
        } else {
            updateArmor();
        }
    }

    @Override
    public void showTitle(Title title) {

    }

    @Override
    public void playSound(Location location, String sound, float volume, float pitch) {

    }

    @Override
    public void initHazardBar(HazardType hazardType, BossBar bar) {

    }

    @Override
    public void setHazardBarProgress(HazardType hazardType, float progress) {

    }

    private void updateArmor() { // updates the armor pieces of the player
        int maxHP = bp.getMaxHP();
        if (visualHp == maxHP) { // hvis playeren har maxa liv så skal de få fullt ut med armor
            mannequin.getEquipment().setArmorContents(new ItemStack[] {
                    getArmorPiece(Material.LEATHER_BOOTS),
                    getArmorPiece(Material.LEATHER_LEGGINGS),
                    getArmorPiece(Material.LEATHER_CHESTPLATE),
                    getArmorPiece(Material.LEATHER_HELMET)});
            return;
        }
        Set<Integer> slots;
        if (maxHP > 5) {
            float d = (float) maxHP / 5;
            int i = (int) Math.ceil((maxHP - visualHp) / d);
            slots = BotBowsPlayer.HEALTH_ARMOR.get(3).get(i - 1);
        } else {
            slots = BotBowsPlayer.HEALTH_ARMOR.get(maxHP - 2).get(maxHP - visualHp - 1);
        }

        for (Integer slot : slots) {
            switch (slot) {
                case 0 -> mannequin.getEquipment().setBoots(null);
                case 1 -> mannequin.getEquipment().setLeggings(null);
                case 2 -> mannequin.getEquipment().setChestplate(null);
                case 3 -> mannequin.getEquipment().setHelmet(null);
            }
        }
    }

    private ItemStack getArmorPiece(Material material) { // makes armor pieces
        ItemStack armor = new ItemStack(material);
        LeatherArmorMeta meta = (LeatherArmorMeta) armor.getItemMeta();
        assert meta != null;
        meta.setColor(bp.getTeam().dyeColor.getColor());
        armor.setItemMeta(meta);
        return armor;
    }

    private AttributeInstance getRequiredAttribute(Attribute attribute) {
        return Objects.requireNonNull( // it should always exist but if not throw
                mannequin.getAttribute(attribute),
                () -> "Missing attribute " + attribute + " for mannequin " + mannequin.getName()
        );
    }
}

File: src/main/java/gruvexp/bbminigames/twtClassic/avatar/NpcAvatar.kt
package gruvexp.bbminigames.twtClassic.avatar

import gruvexp.bbminigames.Main
import gruvexp.bbminigames.twtClassic.BotBows
import gruvexp.bbminigames.twtClassic.BotBowsPlayer
import gruvexp.bbminigames.twtClassic.avatar.BotBowsAvatar.ArmorSet
import gruvexp.bbminigames.twtClassic.effect.PlayerEffectManager
import gruvexp.bbminigames.twtClassic.hazard.HazardType
import net.kyori.adventure.bossbar.BossBar
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextDecoration
import net.kyori.adventure.title.Title
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.attribute.Attribute
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Mannequin
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.LeatherArmorMeta
import org.bukkit.inventory.meta.SkullMeta
import org.bukkit.persistence.PersistentDataType
import org.bukkit.potion.PotionEffect
import java.util.UUID
import kotlin.math.ceil

class NpcAvatar : BotBowsAvatar {
    private var mannequin: Mannequin
    private val _bp: BotBowsPlayer
    private var visualHp = -1
    override lateinit var teamManager: TeamManager

    constructor(mannequin: Mannequin, bp: BotBowsPlayer) {
        this.mannequin = mannequin
        _bp = bp
    }

    constructor(mannequin: Mannequin, previousAvatar: BotBowsAvatar) {
        this.mannequin = mannequin
        _bp = previousAvatar.bp
        teamManager = previousAvatar.teamManager
    }

    override fun message(component: Component) {
    }

    override val entity: LivingEntity
        get() =  mannequin
    override val bp: BotBowsPlayer
        get() = _bp

    override fun eliminate() {
        val newMannequin = Main.WORLD.spawn(bp.team.tribunePos, Mannequin::class.java).apply {
            profile = mannequin.profile
            customName(mannequin.customName())
            isCustomNameVisible = true
        }
        val oldId = uuid
        val newId = newMannequin.uniqueId
        BotBows.replacePlayerId(oldId, newId)
        mannequin.health = 0.0
        mannequin = newMannequin
        setInvis(true)
    }

    override fun revive() {
        setInvis(false)
    }

    override fun setHP(hp: Int) {
        visualHp = hp
        if (hp != 0) updateArmor()
    }

    override fun setMaxHP(maxHP: Int) {
        setHP(maxHP)
    }

    override val armor: ArmorSet
        get() {
            val armor = mannequin.equipment.armorContents
            return ArmorSet(armor[0], armor[1], armor[2], armor[3])
        }

    override fun equipFullArmor() {
        mannequin.equipment.armorContents = arrayOf(
            makeArmorPiece(Material.LEATHER_BOOTS),
            makeArmorPiece(Material.LEATHER_LEGGINGS),
            makeArmorPiece(Material.LEATHER_CHESTPLATE),
            makeArmorPiece(Material.LEATHER_HELMET)
        )
    }

    override fun destroy() {
        mannequin.remove()
    }

    override fun reset() {
    }

    override fun readyBattle(teamManager: TeamManager) {
        this.teamManager = teamManager
    }

    override fun setReady(ready: Boolean, itemIndex: Int) {
    }

    override val nextFreeSlot = 0

    override fun damage() {
        mannequin.damage(0.001)
        bp.effectManager.applyGlow(
            PlayerEffectManager.GlowSource.HIT_COOLDOWN,
            BotBows.HIT_DISABLED_ITEM_TICKS.toLong()
        )
        mannequin.isInvulnerable = true

        Bukkit.getScheduler().runTaskLater(
            Main.getPlugin(),
            Runnable { mannequin.isInvulnerable = false },
            BotBows.HIT_DISABLED_ITEM_TICKS.toLong()
        )
    }

    override var scale: Double
        get() = mannequin.getAttribute(Attribute.SCALE)!!.baseValue
        set(value) {
            mannequin.getAttribute(Attribute.SCALE)!!.baseValue = value
        }

    override fun setGlowing(flag: Boolean) {
        mannequin.isGlowing = flag
    }

    override fun addPotionEffect(effect: PotionEffect) {
        mannequin.addPotionEffect(effect)
    }

    override fun setColor(color: NamedTextColor) {
        teamManager.setColor(mannequin, color)
    }

    override val uuid: UUID
        get() = mannequin.uniqueId

    override val isSneaking
        get() = mannequin.isSneaking

    override fun updateSneakStamina(progress: Float) {
        if (progress >= 1) mannequin.isSneaking = false
    }

    override val headItem: ItemStack
        get() {
            val item = ItemStack(Material.PLAYER_HEAD)
            item.editMeta(SkullMeta::class.java) {
                it.displayName(bp.name.decoration(TextDecoration.ITALIC, false))
                val key = NamespacedKey(Main.getPlugin(), "uuid")
                it.persistentDataContainer.set(key, PersistentDataType.STRING, "${mannequin.uniqueId}")
                it.owningPlayer = Bukkit.getOfflinePlayer("Robotagz")
            }
            return item
        }

    override fun setItem(index: Int, item: ItemStack?) {
    }

    fun setInvis(invis: Boolean) {
        if (invis) {
            mannequin.equipment.clear()
        } else {
            updateArmor()
        }
    }

    override fun showTitle(title: Title) {
    }

    override fun showTitle(component: Component, seconds: Int) {
    }

    override fun playSound(location: Location, sound: String, volume: Float, pitch: Float) {
    }

    override fun initHazardBar(hazardType: HazardType, bar: BossBar) {
    }

    override fun setHazardBarProgress(hazardType: HazardType, progress: Float) {
    }

    private fun updateArmor() { // updates the armor pieces of the player
        val maxHP = bp.settings.maxHealth
        if (visualHp == maxHP) {
            equipFullArmor()
            return
        }
        val slots: Set<Int>
        if (maxHP > 5) {
            val d = maxHP / 5.0
            val i = ceil((maxHP - visualHp) / d).toInt()
            slots = BotBowsPlayer.HEALTH_ARMOR[3][i - 1]
        } else {
            slots = BotBowsPlayer.HEALTH_ARMOR[maxHP - 2][maxHP - visualHp - 1]
        }

        for (slot in slots) {
            when (slot) {
                0 -> mannequin.equipment.setBoots(null)
                1 -> mannequin.equipment.setLeggings(null)
                2 -> mannequin.equipment.setChestplate(null)
                3 -> mannequin.equipment.setHelmet(null)
            }
        }
    }

    private fun makeArmorPiece(material: Material): ItemStack {
        val armor = ItemStack(material)
        armor.editMeta(LeatherArmorMeta::class.java) { it.setColor(bp.team.dyeColor.color) }
        return armor
    }
}
